In [14]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime

In [3]:
#df = pd.read_csv('US_Accidents_Dec19.csv',index_col = 'Start_Time', parse_dates = True)
#df = pd.read_csv('US_Accidents_Dec19.csv',parse_dates = ['Start_Time']).sort_values('Start_Time')
df = pd.read_csv('US_Accidents_Dec19.csv')
 

# using a loop to read more than 1 file
# filenames = ['US_Accidents_Dec19.csv', 'US_Accidents_Dec18.csv']
# dataframes = []
# for f in filenames:
    # dataframes.append(pd.read_csv(f))
    
# or using a comprehension with glob to avoid spelling mistakes
# from glob import blog
# filenames = glob('Accid*.csv')
# dataframes = [pd.read_csv(f) for f in filenames]

In [10]:
df.head()

ID    Source    TMC  Severity           Start_Time             End_Time  \
0  A-1  MapQuest  201.0         3  2016-02-08 05:46:00  2016-02-08 11:00:00   
1  A-2  MapQuest  201.0         2  2016-02-08 06:07:59  2016-02-08 06:37:59   
2  A-3  MapQuest  201.0         2  2016-02-08 06:49:27  2016-02-08 07:19:27   
3  A-4  MapQuest  201.0         3  2016-02-08 07:23:34  2016-02-08 07:53:34   
4  A-5  MapQuest  201.0         2  2016-02-08 07:39:07  2016-02-08 08:09:07   

   Start_Lat  Start_Lng  End_Lat  End_Lng  ...  Roundabout Station   Stop  \
0  39.865147 -84.058723      NaN      NaN  ...       False   False  False   
1  39.928059 -82.831184      NaN      NaN  ...       False   False  False   
2  39.063148 -84.032608      NaN      NaN  ...       False   False  False   
3  39.747753 -84.205582      NaN      NaN  ...       False   False  False   
4  39.627781 -84.188354      NaN      NaN  ...       False   False  False   

  Traffic_Calming Traffic_Signal Turning_Loop Sunrise_Sunset Civil_Twilight  \
0           False          False        False          Night          Night   
1           False          False        False          Night          Night   
2           False           True        False          Night          Night   
3           False          False        False          Night            Day   
4           False           True        False            Day            Day   

  Nautical_Twilight Astronomical_Twilight  
0             Night                 Night  
1             Night                   Day  
2               Day                   Day  
3               Day                   Day  
4               Day                   Day  

[5 rows x 49 columns]

In [9]:
print('Number of rows and columns: ',df.shape)
print('\nFeatures or Column names: \n',df.columns.tolist())
print('\nMissing values: ',df.isnull().values.sum())
#print('\nUnique values: \n',df.nunique())

Number of rows and columns:  (2974335, 49)

Features or Column names: 
 ['ID', 'Source', 'TMC', 'Severity', 'Start_Time', 'End_Time', 'Start_Lat', 'Start_Lng', 'End_Lat', 'End_Lng', 'Distance(mi)', 'Description', 'Number', 'Street', 'Side', 'City', 'County', 'State', 'Zipcode', 'Country', 'Timezone', 'Airport_Code', 'Weather_Timestamp', 'Temperature(F)', 'Wind_Chill(F)', 'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Direction', 'Wind_Speed(mph)', 'Precipitation(in)', 'Weather_Condition', 'Amenity', 'Bump', 'Crossing', 'Give_Way', 'Junction', 'No_Exit', 'Railway', 'Roundabout', 'Station', 'Stop', 'Traffic_Calming', 'Traffic_Signal', 'Turning_Loop', 'Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight', 'Astronomical_Twilight']

Missing values:  11817022


In [18]:
#df.dropna(axis=0, inplace=True)
df.isnull().any().sum()

23

In [6]:
df.shape

(2974335, 49)

In [19]:
df_cond = df[['Severity','State','Weather_Condition','Sunrise_Sunset','Zipcode', 'Start_Time']].head()
df_cond.head(10) #tail

Severity State Weather_Condition Sunrise_Sunset     Zipcode  \
0         3    OH        Light Rain          Night       45424   
1         2    OH        Light Rain          Night  43068-3402   
2         2    OH          Overcast          Night       45176   
3         3    OH     Mostly Cloudy          Night       45417   
4         2    OH     Mostly Cloudy            Day       45459   

            Start_Time  
0  2016-02-08 05:46:00  
1  2016-02-08 06:07:59  
2  2016-02-08 06:49:27  
3  2016-02-08 07:23:34  
4  2016-02-08 07:39:07

In [8]:
#df_cond.info()
#df_cond.describe()

In [22]:
df_loc = df[['State','Zipcode','City']].head()
df_loc.head()

State     Zipcode          City
0    OH       45424        Dayton
1    OH  43068-3402  Reynoldsburg
2    OH       45176  Williamsburg
3    OH       45417        Dayton
4    OH       45459        Dayton

In [23]:
print('Location df shape: \n',df_loc.shape, '\n\nCondition df shape: \n', df_cond.shape)

Location df shape: 
 (5, 3) 

Condition df shape: 
 (5, 6)


### print('Location df shape: \n',df_loc.shape, '\n\nCondition df shape: \n', df_cond.shape)
#### before dropna

Location df shape: 
 (2974335, 7) 

Condition df shape: 
 (2974335, 7)

In [24]:
print(df_loc.isnull().any())
print(df_cond.isnull().any())

State      False
Zipcode    False
City       False
dtype: bool
Severity             False
State                False
Weather_Condition    False
Sunrise_Sunset       False
Zipcode              False
Start_Time           False
dtype: bool


In [11]:
df_loc.dropna(inplace=True)
df_cond.dropna(inplace=True)

/Users/emrentero/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/Users/emrentero/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [13]:
#df_loc['State'].unique()

In [14]:
# merging df_cond and df_loc intro one linking city names and severity to by the zipcode
# df2 = pd.merge(df_cond, df_loc) inner join by default. All columns common to bth DF
# the rows of the merged DF consist of all rows where the columns are identical in both DF cond and loc. 
# df2 = pd.merge(df_cond, df_time, on='Zipcode')  <- this will fixt the problem when rows are different but columns are ==
#In the last line hat will produce duplicates, because columns Zipcode and State are in both DF, so:
# on=['Zipcode', 'State']
# suffixes?
# what if column label doesnt match? EASY!! (BUT! the data on them is similar, and same type)
# pd.merge(df_conc, df_loc, left_on='Zipcode', right_on='State', how='inner', 'outer','left').sorted_values('')
# pd.merge_ordered() when they have a natural order like datetime 

In [15]:
#pd_test = pd.merge(df_loc, df_cond, on=['City'])

In [29]:
df2_merged = pd.merge(df_cond, df_loc, on=['Zipcode', 'State'], how='left')
df2_merged.shape
df2_merged.head()

Severity State Weather_Condition Sunrise_Sunset     Zipcode  \
0         3    OH        Light Rain          Night       45424   
1         2    OH        Light Rain          Night  43068-3402   
2         2    OH          Overcast          Night       45176   
3         3    OH     Mostly Cloudy          Night       45417   
4         2    OH     Mostly Cloudy            Day       45459   

            Start_Time          City  
0  2016-02-08 05:46:00        Dayton  
1  2016-02-08 06:07:59  Reynoldsburg  
2  2016-02-08 06:49:27  Williamsburg  
3  2016-02-08 07:23:34        Dayton  
4  2016-02-08 07:39:07        Dayton

In [33]:
df3_ord_merg = pd.merge_ordered(df_cond, df_loc, on =['Zipcode'], how='outer') #outer by default
df3_ord_merg.shape
df3_ord_merg.head()

Severity State_x Weather_Condition Sunrise_Sunset     Zipcode  \
0         2      OH        Light Rain          Night  43068-3402   
1         2      OH          Overcast          Night       45176   
2         3      OH     Mostly Cloudy          Night       45417   
3         3      OH        Light Rain          Night       45424   
4         2      OH     Mostly Cloudy            Day       45459   

            Start_Time State_y          City  
0  2016-02-08 06:07:59      OH  Reynoldsburg  
1  2016-02-08 06:49:27      OH  Williamsburg  
2  2016-02-08 07:23:34      OH        Dayton  
3  2016-02-08 05:46:00      OH        Dayton  
4  2016-02-08 07:39:07      OH        Dayton

In [53]:
accidents = df2_merged.pivot_table(index= 'Start_Time', values= 'City', columns= ['Sunrise_Sunset','Weather_Condition'], aggfunc='count')
accidents

Sunrise_Sunset                Day      Night                       
Weather_Condition   Mostly Cloudy Light Rain Mostly Cloudy Overcast
Start_Time                                                         
2016-02-08 05:46:00           NaN        1.0           NaN      NaN
2016-02-08 06:07:59           NaN        1.0           NaN      NaN
2016-02-08 06:49:27           NaN        NaN           NaN      1.0
2016-02-08 07:23:34           NaN        NaN           1.0      NaN
2016-02-08 07:39:07           1.0        NaN           NaN      NaN

In [52]:
accidents = accidents.fillna('x')
accidents

Sunrise_Sunset      Day Night
Start_Time                   
2016-02-08 05:46:00   x     1
2016-02-08 06:07:59   x     1
2016-02-08 06:49:27   x     1
2016-02-08 07:23:34   x     1
2016-02-08 07:39:07   1     x

In [ ]:
#PLOTS